File: tests/benchmark/safe/go/parameterized_query.go
// BENCHMARK: safe - parameterized_query
package main

import (
	"database/sql"
	"fmt"
	"net/http"

	_ "github.com/lib/pq"
)

var db *sql.DB

func searchHandler(w http.ResponseWriter, r *http.Request) {
	userID := r.URL.Query().Get("id")
	// Safe: parameterized query with ? placeholder
	rows, err := db.Query("SELECT * FROM users WHERE id = ?", userID)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	defer rows.Close()
	fmt.Fprintf(w, "found rows: %v", rows)
}

func main() {
	http.HandleFunc("/search", searchHandler)
	http.ListenAndServe(":8080", nil)
}

File: tests/benchmark/safe/javascript/parameterized_query.js
// BENCHMARK: safe - parameterized_query
const express = require('express');
const { Pool } = require('pg');
const app = express();
const pool = new Pool();

app.use(express.json());

app.post('/login', async (req, res) => {
    const { username } = req.body;
    // Safe: parameterized query with $1 placeholder
    const result = await pool.query(
        'SELECT * FROM users WHERE username = $1',
        [username]
    );
    res.json(result.rows);
});

app.listen(3000);
